src/storage/workspace_storage.rs

```diff
@@ -266,16 +262,24 @@ impl WorkspaceStorage {
         })
     }
 
-    // Pyo3 is inefficient with Vec<u8> but set_chunk must handle PyByteArray and PyBytes
-    // but can't safely call unsafe { block.as_bytes() } because it's called in multiple threads
-    fn set_chunk(&self, chunk_id: ChunkID, block: Vec<u8>) -> FutureIntoCoroutine {
+    fn set_chunk(
+        &self,
+        py: Python,
+        chunk_id: ChunkID,
+        block: PyObject,
+    ) -> PyResult<FutureIntoCoroutine> {
+        let block = match block.extract::<&PyByteArray>(py) {
+            Ok(x) => x.to_vec(),
+            Err(_) => block.extract::<&PyBytes>(py)?.as_bytes().to_vec(),
+        };
```

Comment:
This should be done using `BytesWrapper` et `unwrap_bytes!` instead

Reply:
Would be done when I rebase the base branch with master